### PR TITLE
Make volumes accessible on SELinux systems by adding relabel option.

### PR DESCRIPTION
**Pull request by tmds**
The volumes mounted by the plugin are not accessible in the container on SELinux systems like Fedora and RHEL.
They need to be relabeled.

For more info see the _Labeling Volume Mounts_ section in https://docs.podman.io/en/latest/markdown/podman-run.1.html.

@bwateratmsft ptal.

### CODE REVIEW

**Comment by bwateratmsft** [comment]
I'm reluctant to have this flag added by default to all volume mounts in all `docker-run` tasks, especially given how ominously the [docs warn](https://docs.docker.com/storage/bind-mounts/#configure-the-selinux-label) about the `z` and `Z` options ("Use **extreme** caution...", emphasis theirs). An alternative option would be to add `rw,z` and `ro,z` as options [here](https://github.com/microsoft/vscode-docker/blob/main/package.json#L1212-L1215), which would allow end users to specify these flags if they wish. It should be noted, you *can* specify those values today, it will work but it will give a yellow squiggle:

image
image

We have logic to implicitly add several volume mappings as needed for debugging .NET and Python. If there are conflicts; i.e. if the user specifies a volume mapping to the same destination in the container, the user's input will be respected. In the example above I'm taking advantage of that logic to get `/src` mapped with `rw,z`.

Do you think updating package.json with `rw,z` and `ro,z` as enum options is sufficient? Or is this likely a narrow enough scenario that the yellow squiggles can just be ignored, with no changes to the extension?

**Comment by tmds** [comment]
> We have logic to implicitly add several volume mappings as needed for debugging .NET and Python.

I'm unable to debug .NET applications because the volumes used by the plugin don't set the `z` flag.
More changes are needed besides updating `package.json` to allow the 'z' option.

> especially given how ominously the docs warn about the z and Z options

I'll look into this a bit deeper and get back to you.

**Comment by bwateratmsft** [comment]
In addition to the changes to `package.json`, you'll need to update your `tasks.json` to override the default volume mappings. Something like this:

```json
        {
            "type": "docker-run",
            "label": "docker-run: debug",
            "dependsOn": [
                "docker-build: debug"
            ],
            "dockerRun": {
                "volumes": [
                    {
                        "containerPath": "/src",
                        "localPath": "${workspaceFolder}",
                        "permissions": "rw,z"
                    },
                    {
                        "containerPath": "/app",
                        "localPath": "${workspaceFolder}",
                        "permissions": "rw,z"
                    },
                    {
                        "containerPath": "/remote_debugger",
                        "localPath": "~/.vsdbg",
                        "permissions": "ro,z"
                    },
                    {
                        "containerPath": "/root/.nuget/packages",
                        "localPath": "~/.nuget/packages",
                        "permissions": "ro,z"
                    },
                ]
            },
            "netCore": {
                "appProject": "${workspaceFolder}/Net6.csproj",
                "enableDebugging": true
            }
        },
```

This should work today albeit with the warning squiggles. For a Windows host, just replace the `~` in the `localPath`'s with `%USERPROFILE%`.

**Comment by tmds** [comment]
> I'll look into this a bit deeper and get back to you.

The comment in the docker docs is out-of-date with the implementation. It will no accept any of the paths mentioned for relabeling.

Trying to relabel them will cause docker to fail.

That means it is not a good idea to unconditionally add the 'z' as the implementation is currently doing.

> We have logic to implicitly add several volume mappings as needed for debugging .NET and Python.

We should make a change that the volumes used for debugging are relabeled. Then they work out of the box on systems with SELinux.

@bwateratmsft do you have a suggestion on how to change `DockerContainerVolume` to support the relabling?

```ts
export interface DockerContainerVolume {
    localPath: string;
    containerPath: string;
    permissions?: 'ro' | 'rw';
}
```

**Comment by bwateratmsft** [comment]
Should we set it so that that flag is only added when `enableDebugging` is set to true? I suppose it's sorta moot since all these implicit volumes are only added if it's true.

For `DockerContainerVolume` you could just add values `ro,z` and `rw,z`, same as for `package.json`.

**Comment by tmds** [comment]
@bwateratmsft I've added the `,z` flag to the debug volumes.

File: src/tasks/netcore/NetCoreTaskHelper.ts [hunk 1]
```diff
@@ -156,6 +156,11 @@ export class NetCoreTaskHelper implements TaskHelper {
 
         runOptions.volumes = await this.inferVolumes(context.folder, runOptions, helperOptions, ssl, userSecrets); // Volumes specifically are unioned with the user input (their input does not override except where the container path is the same)
 
+        // Use a command that keeps the container running when started with the '-t' option.
+        if (runOptions.os !== 'Windows') {
+            runOptions.command = runOptions.command || '/bin/sh';
+        }
```
Comment (tmds) [inline src/tasks/netcore/NetCoreTaskHelper.ts:162]: This is an additional change.

To test this, I had used an image that by default prints usage instructions. This caused the debug session to fail because it assumed the default command would keep the container running.

This explicitly specifies a default command so it has the desired behavior.

@bwateratmsft what do you think about this change?

Comment (bwateratmsft) [reply src/tasks/netcore/NetCoreTaskHelper.ts:162]: My gut feeling is that it should be fine. We do rely on that behavior (of the container staying up and idle); I can't see it being harmful to make it explicit instead of assuming the base images (or in this case, the start command) remain unchanged. Similarly, in Visual Studio, the containers get started with `tail -f /dev/null` as the do-nothing command.

@karolz-ms or @philliphoff can you think of any issues that might arise?

File: src/tasks/DockerRunTaskDefinitionBase.ts [hunk 2]
```diff
@@ -21,7 +21,7 @@ export interface DockerContainerPort {
 export interface DockerContainerVolume {
     localPath: string;
     containerPath: string;
-    permissions?: 'ro' | 'rw';
+    permissions?: 'ro' | 'rw' | 'ro,z' | 'rw,z';
```
Comment (bwateratmsft) [inline src/tasks/DockerRunTaskDefinitionBase.ts:24]: Should we add the `ro,z` and `rw,z` options to the [enum](https://github.com/microsoft/vscode-docker/blob/main/package.json#L1212-L1215) in `package.json` and maybe also the [description string](https://github.com/microsoft/vscode-docker/blob/main/package.nls.json#L85) in `package.nls.json`? This would enable end users to use these options without the warning squiggles.

File: src/tasks/netcore/NetCoreTaskHelper.ts (hunk 1 above)
Comment (karolz-ms) [reply src/tasks/netcore/NetCoreTaskHelper.ts:162]: Hmm. I am not sure about this change. It means that, without additional information, VS Code will not use default command specified by the Docker image. This violates principle of least surprise IMO. And if the image requires a nonstandard command to be "useful", well, we already have a `command` property for that on the `docker-run` task.

Comment (bwateratmsft) [reply src/tasks/netcore/NetCoreTaskHelper.ts:162]: The alternative way to make it explicit would be to scaffold it in, but I don't want to stuff the tasks with so much stuff that it gets overwhelming. I'd worry that scaffolding something like that would cause more problems/confusion than it would solve.

File: src/tasks/DockerRunTaskDefinitionBase.ts (hunk 2 above)
Comment (tmds) [reply src/tasks/DockerRunTaskDefinitionBase.ts]: I have added this. ptal.

Comment (bwateratmsft) [reply src/tasks/DockerRunTaskDefinitionBase.ts]: It's in `package.json` now but got removed from `DockerRunTaskDefinitionBase.ts`, was that intentional? It would need to be in both.

Comment (bwateratmsft) [reply src/tasks/DockerRunTaskDefinitionBase.ts]: Oh, also the changes to `NetCoreTaskHelper.ts` with the `,z` label on each volume mapping got removed.

**Comment by tmds** [comment]
@bwateratmsft I had reverted the wrong commit. I've fixed it now.

**Review by bwateratmsft — APPROVED** [review]
Looks good to me!

File: src/tasks/netcore/NetCoreTaskHelper.ts [hunk 3]
```diff
@@ -208,25 +208,25 @@ export class NetCoreTaskHelper implements TaskHelper {
             const appVolume: DockerContainerVolume = {
                 localPath: path.dirname(helperOptions.appProject),
                 containerPath: runOptions.os === 'Windows' ? 'C:\\app' : '/app',
-                permissions: 'rw'
+                permissions: 'rw,z'
```
Comment (bwateratmsft) [inline src/tasks/netcore/NetCoreTaskHelper.ts:211]: I found one last issue; if Windows containers are used, the `,z` option is not valid. Can you add checks so that `,z` is only added for not-Windows containers? e.g.:
```suggestion
                permissions: runOptions.os === 'Windows' ? 'rw' : 'rw,z'
```

**Review by bwateratmsft — CHANGES_REQUESTED** [review]

File: src/tasks/netcore/NetCoreTaskHelper.ts (hunk 3 above)
Comment (bwateratmsft) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: That's a great idea, I like it!

Comment (tmds) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: I've added the filter.

You may have some suggestions about the style/implementation. (I don't write TypeScript often.)

Comment (bwateratmsft) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: Sorry for the slow response, it's been a very busy week. I'm taking today and next week off so I won't be able to get to this until after the US Thanksgiving holiday.

Comment (tmds) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: @bwateratmsft it would be nice if this, and podman support (https://github.com/microsoft/vscode-docker/pull/3306) are part of the next release (`v1.19`).

I'd like to write a blog post about how use this plugin to debug a .NET app running in a container on Fedora.
Fedora uses podman and SELinux.

Comment (tmds) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: > this, and podman support (#3306)

and https://github.com/microsoft/vscode-docker/issues/3281 also.

Comment (bwateratmsft) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: Sorry again for the slow response. Last week was really hectic. I will get to this PR this week; we are planning to release 1.19 tentatively 17 Jan so this will definitely be a part of it.

Comment (tmds) [reply src/tasks/netcore/NetCoreTaskHelper.ts:211]: No problem. As the changes go in, I'll verify they work well on Fedora.

File: src/tasks/DockerRunTaskProvider.ts [hunk 4]
```diff
@@ -99,4 +99,11 @@ export class DockerRunTaskProvider extends DockerTaskProvider {
             .withQuotedArg(runOptions.image)
             .withArgs(runOptions.command);
     }
+
+    private getVolumeOptions(volume: DockerContainerVolume, isWindows: boolean): string {
```
Comment (bwateratmsft) [inline src/tasks/DockerRunTaskProvider.ts:103]: I'd maybe do this; it's a bit hard to follow the nested ternaries:

```typescript
    private getVolumeOptions(volume: DockerContainerVolume, isWindows: boolean): string {
        if (!volume.permissions) {
            return '';
        } else if (!isWindows) {
            return ':' + volume.permissions;
        } else {
            // The 'z' and 'Z' options aren't supported on Windows containers, normalize to simply ro / rw
            switch (volume.permissions as string) {
                case 'ro,Z':
                case 'ro,z':
                    return ':ro';
                case 'rw,Z':
                case 'rw,z':
                    return ':rw';
                default:
                    return ':' + volume.permissions;
            }
        }
    }
```

File: src/tasks/DockerRunTaskDefinitionBase.ts (hunk 2 above)
Comment (bwateratmsft) [inline src/tasks/DockerRunTaskDefinitionBase.ts:24]: Waiting to hear from @philliphoff if the `,z` option is allowed on Mac. I suspect it is since it works on Linux containers on Windows, but I wanna check for sure...

Comment (bwateratmsft) [reply src/tasks/DockerRunTaskDefinitionBase.ts:24]: Yep, it works. 👍

File: src/tasks/DockerRunTaskProvider.ts (hunk 4 above)
Comment (bwateratmsft) [reply src/tasks/DockerRunTaskProvider.ts:103]: I guess there are also some merge conflicts due to the fix for #3281.

Comment (tmds) [reply src/tasks/DockerRunTaskProvider.ts:104]: Thanks. I've applied this.

**Comment by tmds** [comment]
@bwateratmsft I've addressed your feedback. ptal.

**Review by bwateratmsft — APPROVED** [review]
Looks good to me, thanks @tmds!

**Comment by bwateratmsft** [comment]
This is now released in Docker extension version 1.19.0.
